proposals/low-level-struct-improvements.md

```diff
@@ -284,403 +177,340 @@ ref struct RS
     // Okay: this falls into bullet one above. 
     public ref int Prop1 => ref _refField;
 
-    // Error: This is bullet four above and the *ref-safe-to-escape* of `this`
+    // Error: This is bullet four above and the ref-safe-to-escape of `this`
     // in a `struct` is the current method scope.
     public ref int Prop2 => ref _field;
+}
+```
 
-    public RS(int[] array)
-    {
-        _refField = ref array[0];
-    }
-
-    public RS(ref int i)
-    {
-        _refField = ref i;
-    }
+This may seem like an error at first glance but this is a deliberate design point. Again though, this is not a new rule being created by this proposal, it is instead acknowledging the existing rules `Span<T>` behaved by now that developers can declare their own `ref` state.
 
-    public RS CreateRS() => ...;
+Next the rules for ref re-assignment need to be adjusted for the presence of `ref` fields. The primary scenario for ref re-assignment is `ref struct` constructors storing `ref` parameters into `ref` fields. The support will be more general but this is the core scenario. To support this the rules for ref re-assignment will be adjusted to account for `ref` fields as follows:
 
-    public ref int M1(RS rs)
-    {
-        ref int local1 = ref rs.Prop1;
+<a name="rules-ref-re-assignment"></a>
 
-        // Okay: this falls into bullet two above and the *safe-to-escape* of
-        // `rs` is outside the current method scope. Hence the *ref-safe-to-escape*
-        // of `local1` is outside the current method scope.
-        return ref local;
+> For a ref reassignment in the form ...
+> 1. `x.e1 = ref e2`: where `x` is *safe-to-escape* to *calling method* then `e2` must be *ref-safe-to-escape* to the *calling method*
+> 2. `e1 = ref e2`: where `e1` is a local or parameter, the *ref-safe-to-escape* of `e2` must be at least as wide a scope as the *ref-safe-to-escape* of `e1`.
 
-        // Okay: this falls into bullet two above and the *safe-to-escape* of
-        // `rs` is outside the current method scope. Hence the *ref-safe-to-escape*
-        // of `local1` is outside the current method scope.
-        //
-        // In fact in this scenario you can guarantee that the value returned
-        // from Prop1 must exist on the heap. 
-        RS local2 = CreateRS();
-        return ref local2.Prop1;
+That means the desired `Span<T>` constructor works without any extra annotation:
 
-        // Error: the *safe-to-escape* of `local4` here is the current method 
-        // scope by the revised constructor rules. This falls into bullet two 
-        // above and hence based on that allowed scope.
-        int local3 = 42;
-        var local4 = new RS(ref local3);
-        return ref local4.Prop1;
+```c#
+readonly ref struct Span<T>
+{
+    readonly ref T _field;
+    int _length;
```

Comment:
`_length` should probably be `readonly`.

Reply:
yep, good catch